cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -1,37 +1,59 @@
 //! Implementation of the standard x64 ABI.
 
 use alloc::vec::Vec;
+use log::trace;
 use regalloc::{RealReg, Reg, RegClass, Set, SpillSlot, Writable};
+use std::mem;
 
 use crate::ir::{self, types, types::*, ArgumentExtension, StackSlot, Type};
 use crate::isa::{self, x64::inst::*};
 use crate::machinst::*;
 use crate::settings;
+use crate::{CodegenError, CodegenResult};
 
 use args::*;
 
+/// This is the limit for the size of argument and return-value areas on the
+/// stack. We place a reasonable limit here to avoid integer overflow issues
+/// with 32-bit arithmetic: for now, 128 MB.
+static STACK_ARG_RET_SIZE_LIMIT: u64 = 128 * 1024 * 1024;
+
 #[derive(Clone, Debug)]
 enum ABIArg {
-    Reg(RealReg),
-    _Stack,
+    Reg(RealReg, ir::Type),
+    Stack(i64, ir::Type),
 }
 
-#[derive(Clone, Debug)]
-enum ABIRet {
-    Reg(RealReg),
-    _Stack,
+/// X64 ABI information shared between body (callee) and caller.
+struct ABISig {
+    /// Argument locations (regs or stack slots). Stack offsets are relative to
+    /// SP on entry to function.
+    args: Vec<ABIArg>,
+    /// Return-value locations. Stack offsets are relative to the return-area
+    /// pointer.
+    rets: Vec<ABIArg>,
+    /// Space on stack used to store arguments.
+    stack_arg_space: i64,
+    /// Space on stack used to store return values.
+    stack_ret_space: i64,
```

Comment:
Does space mean size or location?

Reply:
Size, in this case (the `aarch64` impl is the same (and as an aside, we hope to factor out the commonalities here soon)).

Reply:
Why is it an `i64` then? A negative size doesn't make sense.

Reply:
Good question; I'll take care of it during the Grand Refactoring.